src/uct/tcp/tcp_sockcm_ep.c

```diff
@@ -87,6 +73,12 @@ ucs_status_t uct_tcp_sockcm_ep_disconnect(uct_ep_h ep, unsigned flags)
               cep->state, uct_tcp_sockcm_cm_ep_peer_addr_str(cep, peer_str,
                                                              UCS_SOCKADDR_STRING_LEN));
 
+    if ((cep->state & UCT_TCP_SOCKCM_EP_FAILED) &&
+        !(cep->state & UCT_TCP_SOCKCM_EP_GOT_DISCONNECT)) {
+        status = UCS_ERR_NOT_CONNECTED;
+        goto out;
+    }
```

Comment:
maybe shouldn't add this: seems UCP tries to call uct_ep_disconnect() when getting some local error for example in ucp_cm_client_connect_progress() calling uct_cm_client_ep_conn_notify()

Reply:
where does it call uct_ep_disconnect?
note that in case uct_cm_client_ep_conn_notify() fails, it first drops the UCP_EP_FLAG_LOCAL_CONNECTED flag and then goes to the out_free_addr label.

Reply:
right, for my example above it's really not calling disconnect. i'd suggest to check other places in UCP to make sure
@evgeny-leksikov wdyt?

Reply:
seems that the only relevant place in ucp is ucp_cm_client_connect_cb() since it's the function that is called before uct sets the UCT_TCP_SOCKCM_EP_FAILED flag. and there are places in this function where the UCP_EP_FLAG_LOCAL_CONNECTED flag is not dropped before handling an error. @evgeny-leksikov can you please clarify why?

Reply:
in case of local error at UCP level in the middle of wireup protocol UCT does know nothing about failure (both local and remote sides). So locally UCP has to call disconnect to abort connection establishing on both sides.
note, `client_connect` and `server_notify` callbacks return `void`, probably return `ucs_status_t` could simplify UCP and move this responsibility to UCT implementations.

Reply:
So AFAIU in case there is an error on uct and the FAILED flag is set then ucp won't call uct_ep_disconnect from these callbacks. And if ucp does call uct_ep_disconnect, then uct didn't set the FAILED flag. So this assertion can stay. Right?

Reply:
yes